src/txrebroadcast.cpp

```diff
@@ -0,0 +1,219 @@
+            // added it to the set, it would probably not get INVed to most
+            // peers due to filterInventoryKnown.
+            continue;
+        } else {
+            // We have rebroadcasted this transaction before, but will try
+            // again now. Record the attempt.
+            auto UpdateRebroadcastEntry = [start_time](RebroadcastEntry& rebroadcast_entry) {
+                rebroadcast_entry.m_last_attempt = start_time;
+                ++rebroadcast_entry.m_count;
+            };
+
+            m_attempt_tracker->modify(entry_it, UpdateRebroadcastEntry);
+        }
+
+        // Add to set of rebroadcast candidates
+        rebroadcast_txs.push_back(TxIds(txid, wtxid));
+    }
+
+    TrimMaxRebroadcast();
+
+    auto delta1 = after_cnb_time - start_time;
+    auto delta2 = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "GetRebroadcastTransactions(): %d us total, %d us spent in CreateNewBlock.\n", delta2.count(), delta1.count());
+    LogPrint(BCLog::NET, "Queued %d transactions for attempted rebroadcast, filtered from %d candidates with cached fee rate of %s.\n", rebroadcast_txs.size(), block_template->block.vtx.size() - 1, m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB));
+
+    for (TxIds ids : rebroadcast_txs) {
+        LogPrint(BCLog::NET, "Attempting to rebroadcast txid: %s, wtxid: %s\n", ids.m_txid.ToString(), ids.m_wtxid.ToString());
+    }
+
+    return rebroadcast_txs;
+};
+
+void TxRebroadcastHandler::CacheMinRebroadcastFee()
+{
+    if (m_chainman.ActiveChainstate().IsInitialBlockDownload()) return;
+
+    // Update stamp of chain tip on cache run
+    m_tip_at_cache_time = m_chainman.ActiveTip();
+
+    // Update cache fee rate
+    auto start_time = GetTime<std::chrono::microseconds>();
+    m_cached_fee_rate = BlockAssembler(m_chainman.ActiveChainstate(), m_mempool, m_chainparams).MinTxFeeRate();
+    auto delta_time = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "Caching minimum fee for rebroadcast to %s, took %d us to calculate.\n", m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB), delta_time.count());
+};
+
+void TxRebroadcastHandler::RemoveFromAttemptTracker(const CTransactionRef& tx) {
+    const auto it = m_attempt_tracker->find(tx->GetWitnessHash());
+    if (it == m_attempt_tracker->end()) return;
+    m_attempt_tracker->erase(it);
+}
+
+void TxRebroadcastHandler::TrimMaxRebroadcast()
+{
+    // Delete any entries that are older than MAX_ENTRY_AGE
+    auto min_age = GetTime<std::chrono::microseconds>() - MAX_ENTRY_AGE;
+
+    while (!m_attempt_tracker->empty()) {
+        auto it = m_attempt_tracker->get<index_by_last_attempt>().begin();
+        if (it->m_last_attempt < min_age) {
+            m_attempt_tracker->get<index_by_last_attempt>().erase(it);
+        } else {
+            break;
+        }
+    }
+
+    // If there are still too many entries, delete the oldest ones
```

Comment:
hm, then what do you propose as the method to ensure old attempts expire from the tracker?

keep in mind that the time is updated every time a transaction is selected to be rebroadcast.

that said, there still could be the case that the transaction is very high fee rate but hasn't been rebroadcast in a long time because it hit the `MAX_REBROADCAST_COUNT`. if this is the concern, we could address that by re-ordering the if / else logic in `GetRebroadcastTransactions` and update the timestamp if a transaction is being considered to rebroadcast, but filtered due to having hit the max. this way they would stay recent and we could continue with the same deletion logic.

Reply:
Do we agree first on the problem solved by `m_attempt_tracker` ? See new comment : https://github.com/bitcoin/bitcoin/pull/21061#discussion_r627615103

What we want to do in `TrimMaxRebroadcast()` is pending on this conversation imo.

Reply:
ok, going to resolve this conversation here in favor of the other comment thread.